src/Helpers/Compress.php

```diff
@@ -0,0 +1,125 @@
+<?php
+
+namespace Dumbo\Helpers;
+
+use Dumbo\Context;
+use GuzzleHttp\Psr7\Utils;
+use Psr\Http\Message\ResponseInterface;
+use Psr\Http\Message\StreamInterface;
+
+class Compress
+{
+    private const STR_REGEX = '/^\s*(?:text\/[^;\s]+|application\/(?:javascript|json|xml|xml-dtd|ecmascript|dart|postscript|rtf|tar|toml|vnd\.dart|vnd\.ms-fontobject|vnd\.ms-opentype|wasm|x-httpd-php|x-javascript|x-ns-proxy-autoconfig|x-sh|x-tar|x-virtualbox-hdd|x-virtualbox-ova|x-virtualbox-ovf|x-virtualbox-vbox|x-virtualbox-vdi|x-virtualbox-vhd|x-virtualbox-vmdk|x-www-form-urlencoded)|font\/(?:otf|ttf)|image\/(?:bmp|vnd\.adobe\.photoshop|vnd\.microsoft\.icon|vnd\.ms-dds|x-icon|x-ms-bmp)|message\/rfc822|model\/gltf-binary|x-shader\/x-fragment|x-shader\/x-vertex|[^;\s]+?\+(?:json|text|xml|yaml))(?:[;\s]|$)/i';
+
+    /**
+     * Create a middleware that compresses the response body using the specified encoding
+     * if the response size is greater than the given threshold.
+     *
+     * @param array{
+     *     threshold?: int,
+     *     allowedEncodings?: string[],
+     *     encoding?: string
+     * } $options Configuration options
+     * @return callable The middleware
+     */
+    public static function compress(array $options = []): callable
+    {
+        $threshold = $options['threshold'] ?? 1024;
+        $allowedEncodings = ['gzip', 'deflate'];
+        $encoding = $options['encoding'] ?? null;
+
+        return self::compressor($threshold, $allowedEncodings, $encoding);
+    }
+
+    /**
+     * Create a middleware that compresses the response body using the specified encoding
+     * if the response size is greater than the given threshold.
+     *
+     * @param int $threshold The minimum response size in bytes to compress
+     * @param array $allowedEncodings The list of allowed encodings
+     * @param ?string $encoding The selected encoding, or null to auto-detect
+     * @return callable The middleware
+     */
+    private static function compressor(int $threshold, array $allowedEncodings, ?string $encoding): callable
+    {
+        return function (Context $ctx, callable $next) use ($threshold, $allowedEncodings, $encoding) {
+            $next($ctx);
+
+            $response = $ctx->getResponse();
+            $contentLength = $response->getHeaderLine('Content-Length');
+
+            if (
+                $response->hasHeader('Content-Encoding') ||
+                $ctx->req->method() === 'HEAD' ||
+                ($contentLength && (int) $contentLength < $threshold) ||
+                !self::shouldCompress($response) ||
+                !self::shouldTransform($response)
+            ) {
+                return $response;
+            }
+
+            $acceptedEncodings = explode(',', $ctx->req->header('Accept-Encoding'));
```

Comment:
```suggestion
            $acceptedEncodings = array_map('trim', explode(',', $ctx->req->header('Accept-Encoding')));
```

We should probably trim any whitespace here, just to be safe when used outside a browser. Using the array means we get to check the array, and not the value passed too. Some clients might send values as `gzip,br`, `br, gzip`, `deflate,gzip`, etc. etc.

Reply:
Make sense!